core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -755,104 +769,115 @@ private[spark] class BlockManager(
-          putBlockInfo.markFailure()
-          logWarning(s"Putting block $blockId failed")
         }
       }
-    }
-    logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
+      logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
 
-    // Either we're storing bytes and we asynchronously started replication, or we're storing
-    // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
-      data match {
-        case ByteBufferValues(bytes) =>
-          if (replicationFuture != null) {
-            Await.ready(replicationFuture, Duration.Inf)
-          }
-        case _ =>
-          val remoteStartTime = System.currentTimeMillis
-          // Serialize the block if not already done
-          if (bytesAfterPut == null) {
-            if (valuesAfterPut == null) {
-              throw new SparkException(
-                "Underlying put returned neither an Iterator nor bytes! This shouldn't happen.")
+      // Either we're storing bytes and we asynchronously started replication, or we're storing
+      // values and need to serialize and replicate them now:
+      if (putLevel.replication > 1) {
+        data match {
+          case ByteBufferValues(bytes) =>
+            if (replicationFuture != null) {
+              Await.ready(replicationFuture, Duration.Inf)
             }
-            bytesAfterPut = dataSerialize(blockId, valuesAfterPut)
-          }
-          replicate(blockId, bytesAfterPut, putLevel)
-          logDebug("Put block %s remotely took %s"
-            .format(blockId, Utils.getUsedTimeMs(remoteStartTime)))
+          case _ =>
+            val remoteStartTime = System.currentTimeMillis
+            // Serialize the block if not already done
+            if (bytesAfterPut == null) {
+              if (valuesAfterPut == null) {
+                throw new SparkException(
+                  "Underlying put returned neither an Iterator nor bytes! This shouldn't happen.")
+              }
+              bytesAfterPut = dataSerialize(blockId, valuesAfterPut)
+            }
+            replicate(blockId, bytesAfterPut, putLevel)
+            logDebug("Put block %s remotely took %s"
+              .format(blockId, Utils.getUsedTimeMs(remoteStartTime)))
+        }
       }
-    }
 
-    BlockManager.dispose(bytesAfterPut)
+      BlockManager.dispose(bytesAfterPut)
+    } finally {
+      // this is to clean up the byte buffer from the *input* (as opposed to the line above, which
+      // disposes the byte buffer that is the *result* of the put).  We might have turned that byte
+      // buffer into an iterator of values, in which case the input ByteBuffer should be disposed.
+      // It will automatically get disposed when we get to the end of the iterator, but in case
+      // there is some exception before we do, this will take care of it
+      resourceCleaner.doCleanup()
```

Comment:
Cleaning up in a `finally` block should always be the default, unless there's a good explanation for not doing it.